smithy-model/src/main/java/software/amazon/smithy/model/knowledge/IdentifierBindingIndex.java

```diff
@@ -30,13 +33,16 @@
 import software.amazon.smithy.model.traits.RequiredTrait;
 import software.amazon.smithy.model.traits.ResourceIdentifierTrait;
 import software.amazon.smithy.utils.Pair;
+import software.amazon.smithy.utils.SetUtils;
 
 /**
  * Index of operation shapes to the identifiers bound to the operation.
  */
 public final class IdentifierBindingIndex implements KnowledgeIndex {
     /** Map of Resource shape ID to a map of Operation shape ID to a map of identifier name to the member name. */
-    private final Map<ShapeId, Map<ShapeId, Map<String, String>>> bindings = new HashMap<>();
+    private final Map<ShapeId, Map<ShapeId, Map<String, String>>> inputBindings = new HashMap<>();
+    private final Map<ShapeId, Map<ShapeId, Map<String, String>>> outputBindings = new HashMap<>();
+    private final SortedSet<String> allIdentifiers = new TreeSet<>();
```

Comment:
Could this just be computed on demand? Or is it heavily used and worth caching?

Reply:
I don't know for sure. It seems like it could be if a single JVM invocation ran validation, then ran any number of transforms that kept requesting the same information. The choice was made though because it was approach already being taken.